topdown/time.go

```diff
@@ -103,16 +104,66 @@ func builtinClock(a ast.Value) (ast.Value, error) {
 }
 
 func builtinWeekday(a ast.Value) (ast.Value, error) {
-	t, err := utcTime(a)
+	t, err := tzTime(a)
 	if err != nil {
 		return nil, err
 	}
 	weekday := t.Weekday().String()
 	return ast.String(weekday), nil
 }
 
-func utcTime(a ast.Value) (time.Time, error) {
-	value, err := builtins.NumberOperand(a, 1)
+func tzTime(a ast.Value) (t time.Time, err error) {
+	var nVal ast.Value
+	loc := time.UTC
+
+	switch va := a.(type) {
+	case ast.Array:
+
+		if len(va) == 0 {
+			return time.Time{}, builtins.NewOperandTypeErr(1, a, "either number (ns) or [number (ns), string (tz)]")
+		}
+
+		nVal, err = builtins.NumberOperand(va[0].Value, 1)
+		if err != nil {
+			return time.Time{}, err
+		}
+
+		if len(va) > 1 {
+			tzVal, err := builtins.StringOperand(va[1].Value, 1)
+			if err != nil {
+				return time.Time{}, err
+			}
+
+			tzName := string(tzVal)
+
+			switch tzName {
+			case "", "UTC":
+				// loc is already UTC
+			case "Local":
+				loc = time.Local
+
+			default:
+				var ok bool
+				loc, ok = tzCache[tzName]
+
+				if !ok {
+					loc, err = time.LoadLocation(tzName)
+					if err != nil {
+						return time.Time{}, err
+					}
+					tzCache[tzName] = loc
```

Comment:
Sorry I didn't notice it before :( 

We probably should be using a mutex around the `tzCache`. If I understand correctly, because it is global, it could be used by parallel evalutions taking place.

Reply:
Good call. I think it would _probably_ be OK, since it's just a single pointer in an array that's being updated, but better to be explicit.

Reply:
Yea, its probably safe, but iirc if the race detection logic catches us doing it then a panic might happen. https://golang.org/doc/go1.6#runtime has some more details on it.